### ISSUE TITLE
Getting started

### ISSUE THREAD

**Issue by MrCustomizer** [wireapp/scala-utils#1]
I was pointed here after I've posted this question on the Android client's github page: https://github.com/wireapp/wire-android/issues/1064

I am now trying to understand the basics of this project, but I'm having a hard time figuring out where to start. My idea was to use this library to login to the wire production server with my own testing wire account. After the first successful login I wanted to start to exploring more advanced features like getting a list of contacts, sending first messages etc. 

But I haven't been able to find a minimal example which illustrates the login procedure. Looking at the test classes, there seems to be a lot more to the login process than just calling a login method with the username and password. There seem to be some tokens and cookies involved and I'm guessing this has something to do with the handling of the different devices which are associated to an account. But I don't understand which data is optional and which data is essential to connect to the server. I have to admit that I don't know a lot of the wire protocol details, yet. Maybe that's part of the problem. So I would be glad, if you could post some minimal example on how to login to the wire server.

Also, if you think I should be reading up some theoretical stuff about the communication protocol first, I would be happy, if you could point me to some suitable documentation.

I'm not looking for someone explaining me how to use this library. I just need a good starting point from where I can start figuring out stuff on my own. At the moment I'm just overwhelmed by the different account, client and authentication classes and I don't know where to look first.

**Comment by MrCustomizer** [comment]
OK, I think I've accomplished my first successful login with this code:

```
    val accountId = AccountId()

    val client = new DefaultLoginClient(new ApacheHTTPAsyncClient(), BackendConfig.ProdBackend)
    val creds = EmailCredentials(Option(EmailAddress("XXX@XXX.com")), Option("XXXXX"));

    client.login(accountId, creds)

    Thread.sleep(5000)
```

Is this the correct way?

**Comment by MrCustomizer** [comment]
Update: Seems, like the LoginSpec test is the best place to start. I will concentrate my efforts on understanding that test and will report back later.

**Comment by MrCustomizer** [comment]
When trying to run the LoginSpec test, I always get an UnsatisfiedLinkError while trying to load the cross compiled cryptobox library. I've spent several hours trying to figure out what the problem is and I've stumbled upon several stackoverflow threads discussing the obviously problematic task of setting up sbt in a way, that it loads jni libraries without problems. There even is a TODO in the file CryptoBoxServic.scala concerning this issue ("TODO fix unmanaged dependency on cryptobox jni").

I would be willing to spend some time trying to fix this. But first, I have to get a workaround to get it working somehow. As this test exists, I'm suggesting there is a way to get it working and that just the proper integration in sbt is the issue here. But I haven't been able to figure out, how to do this. Do you have some advice?

**Comment by MrCustomizer** [comment]
After spending some more time on this, I get why this TODO hasn't been resolved, yet. My initial error was, that I had been trying to use the *.dylib-files, which are currently located in the lib folder of the core-project. As I'm working in Linux, these files cannot be used, because they are compiled for another operating system. After compiling the cryptobox-jni project in Linux and putting the resulting *.so libraries in the java.library.path, everything works fine.

The thing is: Putting these libraries in one of the system folders which are present in java.library.path is not a good solution, if you want to call it a solution at all. As I currently don't want to change anything in the cryptobox-jni project, the only other solution seems to be to adjust the java.library.path in scala-utils. This is not as trivial as it sounds as it seems as java.library.path is an option, which has to be passed to Java's virtual machine itself. This means, a new virtual machine has to be forked by SBT which just doesn't seem to work in my case. 

This issue is kind of transforming into a journal documenting my progress in finding the current issues with this library and trying to fix them. If you don't have a problem with that, I will continue to write down my progress here. Feel free to chime in, if you have some tips or even a solution to the problems mentioned here.

**Comment by MrCustomizer** [comment]
Seems like I have been able to resolve the issue with the native libraries: https://github.com/wireapp/scala-utils/pull/2

**Comment by MrCustomizer** [comment]
I will close this issue now for the following reasons. The main reason being, that I have lost my trust in the Wire source code while investigating this library. In my opinion, this library suffers from a lack of testing coverage, an overly complicated architecture and a lack of documentation. I haven't been able to send or receive a single message even after having invested a significant amount of time. And it seems abandoned as I don't have received a single response to my questions.
Even worse: I think this library does not work flawlessly, as my new device got constantly added three times instead of one time. It seems to me, that this is a bug in the parallel executed code. On the other hand it is completely possible, that I have just been using the API wrong as it is not clear at all, how to use it correctly. There even seem to be some unresolvable circular dependencies in the service classes…

I'm totally aware of the fact that this library is a side project and should therefore not being held as an example for the rest of Wire's source code. Nevertheless, I'm assuming that the people, who have created this library, are also working on the rest of the client's and server's source code. And that's a thought that troubles me. One paradigm of good and secure code should always be: Keep it as simple as possible. From what I've seen so far of this library and the Android client's source code, the opposite seems to be true.

I'm closing this issue as I've decided that I'm probably leaving Wire in search of another, better chat solution. Perhaps the chat network, created by the people of matrix.org in combination with one of the clients (for example Riot), is the better choice. Matrix's server code is fully open source, btw. And they already support federation.

tl;dr: The source code of this library seems to be a mess and I fear that this could be true for the rest of Wire's source code, too. Leaving this project in search for a better alternative. matrix.org and Riot could be the better choice.

**Comment by someoneEsle** [comment]
Hi, sorry for the late reply. As we mentioned in the other issue, this is a discontinued project that one of our developers had started working on some time ago. It is not meant to be finished, up-to-date, or bug-free: we pointed you to this repository because it could be an interesting starting point for you. But the two main repositories that we actively use are wire-android and the sync engine.
If you want, you can start playing with those. It sounds like you are looking for something more stable and complete, so the official repositories that we build our app from would be a better starting point.
